src/lib/FlightTasks/tasks/FlightTaskOrbit.cpp

```diff
@@ -47,51 +47,85 @@ FlightTaskOrbit::FlightTaskOrbit()
 
 bool FlightTaskOrbit::applyCommandParameters(const vehicle_command_s &command)
 {
-	const float &r = command.param3; /**< commanded radius */
-	const float &v = command.param4; /**< commanded velocity */
+	const float &r = command.param3; // commanded radius
+	const float &v = command.param4; // commanded velocity
+
+	if (setRadius(r) && setVelocity(v)) {
+		return FlightTaskManualAltitudeSmooth::applyCommandParameters(command);
+	}
+
+	return false;
+}
+
+bool FlightTaskOrbit::setRadius(const float r)
+{
+	if (math::isInRange(r, _radius_min, _radius_max)) {
+		// radius is more important than velocity for safety
+		if (!checkAcceleration(r, _v, _acceleration_max)) {
+			_v = sqrtf(_acceleration_max * r);
+		}
 
-	if (math::isInRange(r, 5.f, 50.f) && fabs(v) < 10.f) {
 		_r = r;
+		return true;
+	}
+
+	return false;
+}
+
+bool FlightTaskOrbit::setVelocity(const float v)
+{
+	if (fabs(v) < _velocity_max &&
+	    checkAcceleration(_r, v, _acceleration_max)) {
 		_v = v;
-		return FlightTaskManual::applyCommandParameters(command);
+		return true;
 	}
 
 	return false;
 }
 
+bool FlightTaskOrbit::checkAcceleration(float r, float v, float a)
+{
+	return v * v < a * r;
+}
+
 bool FlightTaskOrbit::activate()
 {
-	bool ret = FlightTaskManual::activate();
-	_r = 1.f;
+	bool ret = FlightTaskManualAltitudeSmooth::activate();
+	_r = _radius_min;
 	_v =  0.5f;
-	_z = _position(2);
 	_center = Vector2f(_position.data());
 	_center(0) -= _r;
 	return ret;
 }
 
 bool FlightTaskOrbit::update()
 {
-	_r += _sticks_expo(0) * _deltatime;
-	_r = math::constrain(_r, 1.f, 20.f);
-	_v -= _sticks_expo(1) * _deltatime;
-	_v = math::constrain(_v, -7.f, 7.f);
-	_z += _sticks_expo(2) * _deltatime;
+	// update altitude
+	FlightTaskManualAltitudeSmooth::update();
 
-	_position_setpoint = Vector3f(NAN, NAN, _z);
+	// stick input adjusts parameters within a fixed time frame
+	const float r = _r + _sticks_expo(0) * _deltatime * (_radius_max / 8.f);
+	const float v = _v - _sticks_expo(1) * _deltatime * (_velocity_max / 4.f);
 
-	/* xy velocity to go around in a circle */
+	setRadius(r);
+	setVelocity(v);
+
+	// xy velocity to go around in a circle
 	Vector2f center_to_position = Vector2f(_position.data()) - _center;
 	Vector2f velocity_xy = Vector2f(center_to_position(1), -center_to_position(0));
 	velocity_xy = velocity_xy.unit_or_zero();
 	velocity_xy *= _v;
 
-	/* xy velocity adjustment to stay on the radius distance */
+	// xy velocity adjustment to stay on the radius distance
 	velocity_xy += (_r - center_to_position.norm()) * center_to_position.unit_or_zero();
```

Comment:
need to figure out why there is a bug when setting position setpoint

Reply:
Yes, for sure. I just didn't want to break existing smooth functionality. Once the problem is found the change here will be small.